File: solutions/Java/EvaluateDivisionUnion-Find.java
import java.util.*;

class Solution {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        Map<String, String> parent = new HashMap<>();
        Map<String, Double> ratio = new HashMap<>();
        
        for (int i = 0; i < equations.size(); i++) {
            String a = equations.get(i).get(0);
            String b = equations.get(i).get(1);
            union(a, b, values[i], parent, ratio);
        }
        
        double[] result = new double[queries.size()];
        for (int i = 0; i < queries.size(); i++) {
            String x = queries.get(i).get(0);
            String y = queries.get(i).get(1);
            if (!parent.containsKey(x) || !parent.containsKey(y) || !find(x, parent, ratio).equals(find(y, parent, ratio))) {
                result[i] = -1.0;
            } else {
                result[i] = ratio.get(x) / ratio.get(y);
            }
        }
        return result;
    }
    
    private void union(String a, String b, double val, Map<String, String> parent, Map<String, Double> ratio) {
        if (!parent.containsKey(a)) {
            parent.put(a, a);
            ratio.put(a, 1.0);
        }
        if (!parent.containsKey(b)) {
            parent.put(b, b);
            ratio.put(b, 1.0);
        }
        String pa = find(a, parent, ratio);
        String pb = find(b, parent, ratio);
        parent.put(pa, pb);
        ratio.put(pa, val * ratio.get(b) / ratio.get(a));
    }
    
    private String find(String x, Map<String, String> parent, Map<String, Double> ratio) {
        if (!parent.get(x).equals(x)) {
            String root = find(parent.get(x), parent, ratio);
            ratio.put(x, ratio.get(x) * ratio.get(parent.get(x)));
            parent.put(x, root);
        }
        return parent.get(x);
    }
}

File: solutions/C++/evaluate_division_union-find.cpp
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, pair<string, double>> parent;
        
        for (int i = 0; i < equations.size(); i++) {
            unite(equations[i][0], equations[i][1], values[i], parent);
        }
        
        vector<double> result;
        for (const auto& query : queries) {
            string x = query[0], y = query[1];
            if (parent.find(x) == parent.end() || parent.find(y) == parent.end()) {
                result.push_back(-1.0);
            } else {
                auto [px, rx] = find(x, parent);
                auto [py, ry] = find(y, parent);
                if (px != py) {
                    result.push_back(-1.0);
                } else {
                    result.push_back(rx / ry);
                }
            }
        }
        return result;
    }
    
private:
    void unite(const string& a, const string& b, double val, unordered_map<string, pair<string, double>>& parent) {
        if (parent.find(a) == parent.end()) parent[a] = {a, 1.0};
        if (parent.find(b) == parent.end()) parent[b] = {b, 1.0};
        
        auto [pa, ra] = find(a, parent);
        auto [pb, rb] = find(b, parent);
        parent[pa] = {pb, val * rb / ra};
    }
    
    pair<string, double> find(const string& x, unordered_map<string, pair<string, double>>& parent) {
        if (parent[x].first != x) {
            auto [root, ratio] = find(parent[x].first, parent);
            parent[x].second *= ratio;
            parent[x].first = root;
        }
        return {parent[x].first, parent[x].second};
    }
};
